server/src/main/java/org/elasticsearch/index/codec/storedfields/TSDBStoredFieldsFormat.java

```diff
@@ -0,0 +1,282 @@
+            return storedFieldsWriter.ramBytesUsed() + bloomFilterStoredFieldsWriter.ramBytesUsed();
+        }
+
+        private StoredFieldsWriter getWriterForField(FieldInfo field) {
+            if (field.number == idFieldNumber || field.name.equals(IdFieldMapper.NAME)) {
+                idFieldNumber = field.number;
+                return bloomFilterStoredFieldsWriter;
+            }
+            return storedFieldsWriter;
+        }
+    }
+
+    class TSDBStoredFieldsReader extends StoredFieldsReader implements BloomFilter {
+        private final StoredFieldsReader storedFieldsReader;
+        private final StoredFieldsReader bloomFilterStoredFieldsReader;
+        private final BloomFilter bloomFilter;
+
+        TSDBStoredFieldsReader(Directory directory, SegmentInfo si, FieldInfos fn, IOContext context) throws IOException {
+            boolean success = false;
+            List<Closeable> toClose = new ArrayList<>();
+            try {
+                this.storedFieldsReader = storedFieldsFormat.fieldsReader(directory, si, fn, context);
+                toClose.add(this.storedFieldsReader);
+                this.bloomFilterStoredFieldsReader = bloomFilterStoredFieldsFormat.fieldsReader(directory, si, fn, context);
+                this.bloomFilter = (BloomFilter) bloomFilterStoredFieldsReader;
+                toClose.add(this.bloomFilterStoredFieldsReader);
+                success = true;
+            } finally {
+                if (success == false) {
+                    IOUtils.close(toClose);
+                }
+            }
+        }
+
+        TSDBStoredFieldsReader(StoredFieldsReader storedFieldsReader, StoredFieldsReader bloomFilterStoredFieldsReader) {
+            this.storedFieldsReader = storedFieldsReader;
+            this.bloomFilterStoredFieldsReader = bloomFilterStoredFieldsReader;
+            assert bloomFilterStoredFieldsReader instanceof BloomFilter;
+            this.bloomFilter = (BloomFilter) bloomFilterStoredFieldsReader;
+        }
+
+        @Override
+        public StoredFieldsReader clone() {
+            return new TSDBStoredFieldsReader(storedFieldsReader.clone(), bloomFilterStoredFieldsReader.clone());
+        }
+
+        @Override
+        public StoredFieldsReader getMergeInstance() {
+            return new TSDBStoredFieldsReader(storedFieldsReader.getMergeInstance(), bloomFilterStoredFieldsReader.getMergeInstance());
+        }
+
+        @Override
+        public void checkIntegrity() throws IOException {
+            storedFieldsReader.checkIntegrity();
+            bloomFilterStoredFieldsReader.checkIntegrity();
+        }
+
+        @Override
+        public void close() throws IOException {
+            IOUtils.close(storedFieldsReader, bloomFilterStoredFieldsReader);
+        }
+
+        @Override
+        public void document(int docID, StoredFieldVisitor visitor) throws IOException {
+            bloomFilterStoredFieldsReader.document(docID, visitor);
```

Comment:
Should we add a comment that it is expected to call the bloom filter reader first when visiting a document?